### PR TITLE
enhancement(networking)!: Fibonacci instead of fixed backoff

### PR DESCRIPTION
**Pull request by lukesteensen**
This is just a quick spike to start the discussion, not quite ready to be merged.

We currently use a fixed delay between retries, which is not ideal. Fibonacci backoff is a simple strategy that does [pretty well](https://pdfs.semanticscholar.org/e575/e12ef56bcd889ce09516dd702ec1422816b6.pdf) in evaluations and is easy to implement with no added dependencies.

If we want to switch to this strategy, there are a few things we'll need to decide:

1. Do we change the `retry_backoff_secs` config item to specify that it's a base unit of a more complex strategy?
2. How should we cap the delay? The spike just hardcoded 10 seconds as the max, but this is something that could vary greatly between different users and different downstream systems. Would we add a `max_backoff_secs` config item to every relevant sink?

### CODE REVIEW

**Comment by binarylogic** [comment]
I spoke to @lukesteensen about this offline and mentioned [this AWS post on exponential backoffs + jitter](https://aws.amazon.com/blogs/architecture/exponential-backoff-and-jitter/).

@LucioFranco @a-rodin do either of you have inisghts or experience with this?

**Comment by binarylogic** [comment]
> Do we change the retry_backoff_secs config item to specify that it's a base unit of a more complex strategy?

Also, this might be a good time to introduce #1022. I'd vote for a `retry_policy.strategy` attribute that defaults to `fibonacci`. I think the value is probably pretty low for configurable strategies, but I assume there are scenarios where  another strategy would be preferrable. Ex: [the AWS post](https://aws.amazon.com/blogs/architecture/exponential-backoff-and-jitter/) I previously linked to.

**Comment by a-rodin** [comment]
I don't really have experience with Fibonacci backoff, but I'm thinking could it make sense to provide ability to set not only initial backoff, but ability to skip steps?

My reasoning is that for exponential backoff we have two parameter to control, initial value and the exponent base. For Fibonacci numbers we can set only the initial value. But as an analog for the base it is possible to take not every number from the Fibonacci sequence, but instead only each `n`th number where `n` is a user-controlled parameter

**Comment by lukesteensen** [comment]
> I think the value is probably pretty low for configurable strategies, but I assume there are scenarios where another strategy would be preferable. Ex: [the AWS post](https://aws.amazon.com/blogs/architecture/exponential-backoff-and-jitter/) I previously linked to.

Yeah, the value of supporting a bunch of different strategies is likely very low. The post you linked is specifically about the case of optimistic concurrency control and conflicting updates to a database. I don't really expect that requests from Vector to an external system will ever be considered conflicting and cause correlated failures in the same way. We could do a little simulation of our own, but our model (service being up/down/overloaded) isn't as interesting or precise.

> I don't have experience with Fibonacci backoff, but I'm thinking could it make sense to provide ability to set not only initial backoff, but ability to skip steps?

I see the simpler configuration as a benefit of fibonacci over exponential. We could definitely explore an option like this if we see some demand from users, but I don't expect many people to need that level of control.

**Comment by LucioFranco** [comment]
> But as an analog for the base it is possible to take not every number from the Fibonacci sequence, but instead only each nth number where n is a user-controlled parameter

Good idea 👍 

> Do we change the retry_backoff_secs config item to specify that it's a base unit of a more complex strategy?

what about `initial_backoff_delay_secs`?

> How should we cap the delay? The spike just hardcoded 10 seconds as the max, but this is something that could vary greatly between different users and different downstream systems. Would we add a max_backoff_secs config item to every relevant sink?

I think it makes sense to add that config option. Would we want to default like we do now to max u64?

This overall looks great!

**Comment by binarylogic** [comment]
Noting, we are waiting to merge this since it is a breaking change. We'll be bundling it with other configuration changes.

**Comment by LucioFranco** [comment]
Ok @lukesteensen this has the rest of the changes if you can take a quick look.

@binarylogic I was having trouble adding a new field to `batching_sink.rb`, I am going to push now an update to the descriptions but will need some help changing the field names and adding a new one.

As a summary, the changes are as follows:

- Instead of fixed delays between retries, we use the Fibonacci sequence. 
- `retry_backoff_secs` has been changed to `retry_initial_backoff_secs` since it only describes the first delay from first failed request and first retry. This is also the seed of the fib sequence.
- A new `retry_max_duration_secs` option which specifies the largest delay between retries, current default is set to 10 seconds.

File: scripts/util/metadata/batching_sink.rb [hunk 1]
```diff
@@ -124,7 +124,7 @@ def initialize(hash)
       "retry_backoff_secs" =>
```
Comment (LucioFranco) [inline scripts/util/metadata/batching_sink.rb]: @binarylogic this needs to be renamed to `retry_initial_backoff_secs`

Comment (binarylogic) [reply scripts/util/metadata/batching_sink.rb]: Done

**Pull request merged**